gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,792 @@
+        """Check that the training parameters provided make sense. e.g. raise error if `epochs` not provided"""
+        raise NotImplementedError
+
+    def _set_keyedvectors(self):
+        raise NotImplementedError
+
+    def _worker_loop(self, job_queue, progress_queue):
+        """Train the model, lifting lists of data from the job_queue."""
+        thread_private_mem = self._get_thread_working_mem()
+        jobs_processed = 0
+        while True:
+            job = job_queue.get()
+            if job is None:
+                progress_queue.put(None)
+                break  # no more jobs => quit this worker
+            data_iterable, job_parameters = job
+
+            for callback in self.callbacks:
+                callback.on_batch_begin(self)
+
+            tally, raw_tally = self._do_train_job(data_iterable, job_parameters, thread_private_mem)
+
+            for callback in self.callbacks:
+                callback.on_batch_end(self)
+
+            progress_queue.put((len(data_iterable), tally, raw_tally))  # report back progress
+            jobs_processed += 1
+        logger.debug("worker exiting, processed %i jobs", jobs_processed)
+
+    def _job_producer(self, data_iterator, job_queue, cur_epoch=0, total_examples=None, total_words=None):
+        """Fill jobs queue using the input `sentences` iterator."""
+        job_batch, batch_size = [], 0
+        pushed_words, pushed_examples = 0, 0
+        next_job_params = self._get_job_params(cur_epoch)
+        job_no = 0
+
+        for data_idx, data in enumerate(data_iterator):
+            data_length = self._raw_word_count([data])
+
+            # can we fit this sentence into the existing job batch?
+            if batch_size + data_length <= self.batch_words:
+                # yes => add it to the current job
+                job_batch.append(data)
+                batch_size += data_length
+            else:
+                job_no += 1
+                job_queue.put((job_batch, next_job_params))
+
+                # update the learning rate for the next job
+                if total_examples:
+                    # examples-based decay
+                    pushed_examples += len(job_batch)
+                    epoch_progress = 1.0 * pushed_examples / total_examples
+                else:
+                    # words-based decay
+                    pushed_words += self._raw_word_count(job_batch)
+                    epoch_progress = 1.0 * pushed_words / total_words
+                next_job_params = self._update_job_params(next_job_params, epoch_progress, cur_epoch)
+
+                # add the sentence that didn't fit as the first item of a new job
+                job_batch, batch_size = [data], data_length
+        # add the last job too (may be significantly smaller than batch_words)
+        if job_batch:
+            job_no += 1
+            job_queue.put((job_batch, next_job_params))
+
+        if job_no == 0 and self.train_count == 0:
+            logger.warning(
```

Comment:
What's a reason to use `logger.warning` here (instead of an exception, or, at least `warning.warn`)?

Reply:
Is this new? I believe this is the old code.

Reply:
Yes, this is copy-pasted from old code.